File: src/code/java/graph/dfs_iterative.java
import java.util.Stack;
import java.util.HashSet;
import java.util.Set;


public int fn(int[][] graph) {
    Stack<Integer> stack = new Stack<>();
    Set<Integer> seen = new HashSet<>();
    stack.push(START_NODE);
    seen.add(START_NODE);
    int ans = 0;

    while (!stack.empty()) {
        int node = stack.pop();
        // TODO: logic
        for (int neighbor: graph[node]) {
            if (!seen.contains(neighbor)) {
                seen.add(neighbor);
                stack.push(neighbor);
            }
        }
    }

    return ans;
}

File: src/code/ruby/graph/dfs_iterative.rb
require 'set'

def fn(graph)
  stack = [START_NODE]
  seen  = Set.new([START_NODE])
  ans   = 0

  until stack.empty?
    node = stack.pop
    # TODO: logic
    graph[node].each do |neighbor|
      unless seen.include?(neighbor)
        seen.add(neighbor)
        stack << neighbor
      end
    end
  end

  ans
end
